text/0000-futures.md

```diff
@@ -154,31 +162,325 @@ can be dropped.
+customize the behavior of `RawWaker`, `Waker` and `LocalWaker` objects.
+
+This mechanism is chosen in favor of trait objects since it allows for more
+flexible memory management schemes. `RawWaker` can be implemented purely in
+terms of global functions and state, on top of reference counted objects, or
+in other ways.
+
+The relation between those `Waker` types is outlined in the following definitions:
+
+```rust
+/// A `RawWaker` allows the implementor of a task executor to customize the
+/// behavior of `LocalWaker`s and `Waker`s.
+/// It consists of a data pointer a virtual function pointer table (vtable) that
+/// customizes the behavior of the `RawWaker`.
+#[derive(PartialEq)]
+pub struct RawWaker {
+    /// A data pointer, which can be used to store arbitrary data as required
+    /// by the executor. This could be e.g. a type-erased pointer to an `Arc`
+    /// that is associated with the task.
+    /// The value of this field gets passed to all functions that are part of
+    /// the vtable as first parameter.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior if this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that allows to customize the
+/// behavior of a `RawWaker`.
+/// The pointers which get passed to all functions inside the vtable are the
+/// values of the `data` field.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` or `LocalWaker` in which the `RawWaker` is stored gets cloned.
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a `RawWaker` and associated
+    /// task.
+    pub clone: unsafe fn(*const ()) -> *const (),
+    /// This function will be called when a `LocalWaker` should be converted into
+    /// a thread-safe `Waker`. The implementation of this function must return
+    /// a new `RawWaker` which can fulfill the requirements of `Waker`. It can
+    /// exchange the vtable for that purpose. E.g. it might replace the `wake`
+    /// function with a varient that supports cross-thread wakeups.
+    ///
+    /// The old `LocalWaker` which contained the data pointer should be consumed
+    /// while performing the operation. After the operation the `LocalWaker`
+    /// won't exist anymore, only the new `Waker`.
+    /// This means that if both instances would utilize the same reference-counted
+    /// data, changing the reference count would not be necessary.
+    ///
+    /// If a conversion is not supported, the implementation should return
+    /// `INVALID_RAW_WAKER`. In this case the implementation must still make
+    /// sure that the data pointer which is passed to the function is correctly
+    /// released, e.g. by calling the associated `drop_fn` on it.
+    pub into_waker: unsafe fn(*const ()) -> RawWaker,
```

Comment:
We could consider returning an `Option` instead?

Reply:
Yes, that's actually a lot better. I changed it.